pkg/dnsmasq/deployment.go

```diff
@@ -0,0 +1,192 @@
+	"github.com/openstack-k8s-operators/lib-common/modules/common/affinity"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/env"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	// ServiceCommand -
+	ServiceCommand = "dnsmasq"
+	// DebugCommand -
+	DebugCommand = "/bin/sleep infinity"
+)
+
+// Deployment func
+func Deployment(
+	instance *networkv1.DNSMasq,
+	configHash string,
+	labels map[string]string,
+	annotations map[string]string,
+	cms *corev1.ConfigMapList,
+) *appsv1.Deployment {
+	runAsUser := int64(0)
+	terminationGracePeriodSeconds := int64(10)
+
+	livenessProbe := &corev1.Probe{
+		// TODO might need tuning
+		TimeoutSeconds:      5,
+		PeriodSeconds:       3,
+		InitialDelaySeconds: 3,
+	}
+	readinessProbe := &corev1.Probe{
+		// TODO might need tuning
+		TimeoutSeconds:      5,
+		PeriodSeconds:       5,
+		InitialDelaySeconds: 5,
+	}
+
+	command := []string{"/bin/bash"}
+	args := []string{"-c"}
+	initArgs := []string{"-c"}
+	if instance.Spec.Debug.Service {
+		args = append(args, DebugCommand)
+		initArgs = append(initArgs, "/bin/true")
+
+		livenessProbe.Exec = &corev1.ExecAction{
+			Command: []string{
+				"/bin/true",
+			},
+		}
+
+		readinessProbe.Exec = &corev1.ExecAction{
+			Command: []string{
+				"/bin/true",
+			},
+		}
+	} else {
+		dnsmasqCmd := []string{ServiceCommand}
+		dnsmasqCmd = append(dnsmasqCmd, "--interface=*")
+		dnsmasqCmd = append(dnsmasqCmd, "--conf-dir=/etc/dnsmasq.d")
+		dnsmasqCmd = append(dnsmasqCmd, "--hostsdir=/etc/dnsmasq.d/hosts")
+		dnsmasqCmd = append(dnsmasqCmd, "--keep-in-foreground")
+		dnsmasqCmd = append(dnsmasqCmd, "--no-daemon")
+		dnsmasqCmd = append(dnsmasqCmd, "--log-debug")
+		dnsmasqCmd = append(dnsmasqCmd, "--bind-interfaces")
+		dnsmasqCmd = append(dnsmasqCmd, "--listen-address=$(POD_IP)")
+		dnsmasqCmd = append(dnsmasqCmd, "--port "+strconv.Itoa(int(DNSPort)))
+		// log to stdout
+		dnsmasqCmd = append(dnsmasqCmd, "--log-facility=-")
+		// dns
+		dnsmasqCmd = append(dnsmasqCmd, "--no-hosts")
+		dnsmasqCmd = append(dnsmasqCmd, "--domain-needed")
+		dnsmasqCmd = append(dnsmasqCmd, "--no-resolv")
+		dnsmasqCmd = append(dnsmasqCmd, "--bogus-priv")
+		dnsmasqCmd = append(dnsmasqCmd, "--log-queries")
+
+		// append dnsmasqCmd for service container
+		args = append(args, strings.Join(dnsmasqCmd, " "))
+
+		// append --test for initcontainer check config syntax
+		dnsmasqCmd = append(dnsmasqCmd, "--test")
+		initArgs = append(initArgs, strings.Join(dnsmasqCmd, " "))
+
+		//
+		// https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/
+		//
+		livenessProbe.TCPSocket = &corev1.TCPSocketAction{
+			Port: intstr.IntOrString{Type: intstr.Int, IntVal: int32(DNSPort)},
+		}
+		readinessProbe.TCPSocket = &corev1.TCPSocketAction{
+			Port: intstr.IntOrString{Type: intstr.Int, IntVal: int32(DNSPort)},
+		}
```

Comment:
it listens on TCP and UDP. since you can not test UDP properly, right now I expect that when its up and listens on TCP, UDP works as well. We probably want to look if there is a good way to improve that.